### ISSUE TITLE
URL encoded route parameters not handle correctly

### ISSUE THREAD

**Issue by tomqwpl** [go-chi/chi#641]
There was a fix put in (#148) a while ago that introduced use of url.RawPath as the basis for the route path.
However, as I understand it, RawPath is only a _hint_ and doesn't always include the actual URL encoded path. It only contains the path if it differs from the default encoding of the path. In other words, RawPath being blank doesn't mean that the path requires URL encoding, or that the path originally contained no URL encoded elements, just that the original path can be reconstructed from path.
The relevant code in url is:
```
// setPath sets the Path and RawPath fields of the URL based on the provided
// escaped path p. It maintains the invariant that RawPath is only specified
// when it differs from the default encoding of the path.
// For example:
// - setPath("/foo/bar")   will set Path="/foo/bar" and RawPath=""
// - setPath("/foo%2fbar") will set Path="/foo/bar" and RawPath="/foo%2fbar"
// setPath will return an error only if the provided path contains an invalid
// escaping.
func (u *URL) setPath(p string) error {
	path, err := unescape(p, encodePath)
	if err != nil {
		return err
	}
	u.Path = path
	if escp := escape(path, encodePath); p == escp {
		// Default encoding is fine.
		u.RawPath = ""
	} else {
		u.RawPath = p
	}
	return nil
}
```

Note that the condition is not "does the path need escaping" but "does the original rawpath equal the escaped form of the path".

In other words, rather than consulting RawPath, I believe that the code should be consulting EscapedPath().

There is no particularly easy workaround to this. I can't decode the result from chi.URLParam, because it _might_ already be decoded, but might not.

This appears to be a bug in the fix supplied for #148, or rather that the fix didn't cater for all scenarios. 

I'm happy to provide a PR for this if it's felt that this is a bug.

**Comment by tomqwpl** [comment]
Perhaps this is best explained by a test case. Note that this test case shows what actually happens now:

```

func TestEscapedURLParams2(t *testing.T) {
	m := NewRouter()
	m.Get("/api/{identifier}", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(200)
		rctx := RouteContext(r.Context())
		if rctx == nil {
			t.Error("no context")
			return
		}
		identifier := URLParam(r, "identifier")
		w.Write([]byte(identifier))
	})

	ts := httptest.NewServer(m)
	defer ts.Close()

	if _, body := testRequest(t, ts, "GET", "/api/hello world", nil); body != "hello world" { // 1
		t.Fatalf(body)
	}
	if _, body := testRequest(t, ts, "GET", "/api/hello%20world", nil); body != "hello world" { // 3
		t.Fatalf(body)
	}
	if _, body := testRequest(t, ts, "GET", "/api/hello%2520world", nil); body != "hello%20world" { // 4
		t.Fatalf(body)
	}
	if _, body := testRequest(t, ts, "GET", "/api/hello%2Fworld", nil); body != "hello%2Fworld" { // 5
		t.Fatalf(body)
	}
}
```

If this was a real handler, I wouldn't apparently know whether to unescape the result I get. 
The first two cases show expected unescaping behaviour. In normal cases the parameter is unescaped, which I think is what you expect to happen.
In case 4 I wouldn't want unescape the parameter, as doing so would result in the wrong value. The value chi gives is already the right value. 
In case 5 however I would want to unescape the result since unless I do I don't get the right result. 

So you can see that at the moment it seems that chi gives inconsistent results in terms of whether the strings returned are URL coded or not, and you don't really know whether you need to decode them or not.

**Comment by pkieltyka** [comment]
thanks for pointing out the inconsistency. I agree it should be at least consistent. The choice should be the one which is more efficient and requires less processing on chi's side.

**Comment by tomqwpl** [comment]
Further examples of inconsistency introduced. Again these tests display the observed behaviour. The implementation of url.setPath means that the case in which you URL encode things affects whether the parameter is extracted unescaped or not:
```

	if _, body := testRequest(t, ts, "GET", "/api/%7B%7D", nil); body != "{}" {
		t.Fatalf("%s", body)
	}
	if _, body := testRequest(t, ts, "GET", "/api/%7b%7d", nil); body != "%7b%7d" {
		t.Fatalf("%s", body)
	}
```
Ignoring efficiency, it feels like the right implementation is to always parse the result of url.EscapedPath, and then unescape the individual route parameters having extracted them
Selfishly the behaviour I'm after is to be able to implement a route parameter that is a path, so contains encoded forward slashes.

**Comment by tomqwpl** [comment]
In a fork I've created a fix that makes a simple change that appears to produce a consistent result. The existing tests all work, with the exception of the TestEscapedURLParams test were the test is expecting what is essentially inconsistent behaviour. My additional tests also work.
https://github.com/tomqwpl/chi/commit/22e0f03cbc24e2915242164aa20f511da507a64e

I don't particularly want to maintain a long term fork of chi, so if you don't think that's a good change for chi to make I'll have to work out an alternative plan.

**Comment by tomqwpl** [comment]
Gorilla mux has an option that says "use encoded path". It does the obvious thing. By default it will parse the unescaped path, as chi normally does. Route parameters are then returned to you unescaped.
You can set an option of "use encoded path", in which case all that happens is that it uses url.EscapedPath. Parameters are therefore returned to you escaped, and its up to you to unescape them. This seems like sensible behaviour as it means you have to opt in to the modified behaviour.
The only issue is that the current chi behaviour has this inconsistency, and people may be relying on the way in which it can apparently deal with URL encoded paths. However the existing behaviour also has these subtle inconsistencies where the case in which you URL encode things will affect the result.

Actually, I realise I can get the behaviour I desire using a middleware. It's a bit of a hack, but a middleware that sets "r.URL.RawPath=r.URL.EncodedPath" then exploits the behaviour that chi will parse RawPath if it is present. The parameters then get returned unescaped, but that's fine, as I know it'll always use the encoded path, so I can safely unescape the parameters without risk of double unescaping. It's fragile though, especially if you are considering tidying this up to remove the inconsistencies.

**Comment by bronger** [comment]
Please tell me whether I am affected by this.  I have the following route installed:

```go
r.Get("/{instId:[A-Za-z0-9-():]+}", instPage)
```

Then, the URL `/I:%28x%29` is correctly routed and the result of `chi.URLParam` is `I:(x)`, which is correct. However, if I encode the colon with `%3a`, all breaks down: The request is not routed to `instPage`, and if I make the URL pattern more general to re-enable the routing, `chi.URLParam` returns a totally escaped `I%3a%28x%29`.

Is this the same issue?
